enclave/src/openfinex/fixed_point_number_converter.rs

```diff
@@ -0,0 +1,256 @@
+}
+
+fn combine_fraction_digits(digits: &Vec<u8>) -> OpenFinexApiResult<ResponseInteger> {
+    combine_digits(digits, &|order| (UNIT / 10u128).checked_div(order))
+}
+
+fn combine_digits(
+    digits: &Vec<u8>,
+    scale_fn: &dyn Fn(u128) -> Option<u128>,
+) -> OpenFinexApiResult<ResponseInteger> {
+    let mut number: u128 = 0;
+    let mut order: u128 = 1;
+
+    for digit in digits {
+        // all operations are performed in 'checked' mode to prevent overflow errors
+
+        let scale = (scale_fn)(order).ok_or_else(|| {
+            OpenFinexApiError::FixedPointConversionError("Value overflow".to_string())
+        })?;
+
+        let digit_value = scale.checked_mul(*digit as u128).ok_or_else(|| {
+            OpenFinexApiError::FixedPointConversionError("Value overflow".to_string())
+        })?;
+
+        number = number.checked_add(digit_value).ok_or_else(|| {
+            OpenFinexApiError::FixedPointConversionError("Value overflow".to_string())
+        })?;
+
+        order = order.checked_mul(10u128).ok_or_else(|| {
+            OpenFinexApiError::FixedPointConversionError("Value overflow".to_string())
+        })?;
+    }
+
+    Ok(number)
+}
+
+pub mod tests {
+
+    use super::*;
+
+    pub fn fail_to_parse_invalid_strings() {
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("ar4")).is_err());
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("34.5g")).is_err());
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("not_a_number")).is_err());
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("12.f")).is_err());
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("a.56")).is_err());
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("a0.21542")).is_err());
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("")).is_err());
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("_")).is_err());
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("-1547566.2894")).is_err());
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("NaN")).is_err());
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("Inf")).is_err());
+    }
+
+    pub fn fail_to_parse_scientific_notation() {
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("1.4e-3")).is_err());
+        assert!(FixedPointNumberConverter::parse_from_string(&format!("9e8")).is_err());
+    }
+
+    pub fn fail_to_parse_if_too_large() {
+        // number in string is larger than what can be converted to u128 (with 18 digit shift)
+        // u128 max 340_282_366_920_938_463_463,,_374_607_431_768_211_455
+        assert!(
+            FixedPointNumberConverter::parse_from_string(&format!("340282366920938463464"))
+                .is_err()
+        );
+    }
+
+    pub fn successfully_parse_numbers() {
```

Comment:
Below are examples of numbers that can be parsed successfully and their corresponding integer value